src/finalfusion/compat/text.py

```diff
@@ -0,0 +1,149 @@
+    with open(file) as inf:
+        rows, cols = next(inf).split()
+        return _load_text(inf, int(rows), int(cols))
+
+
+def load_text(file: Union[str, bytes, int, PathLike]) -> Embeddings:
+    """
+    Read embeddings in text format.
+
+    The returned embeddings have a SimpleVocab, NdArray storage and a Norms chunk. The storage is
+    l2-normalized per default and the corresponding norms are stored in the Norms.
+
+    Expects a file with utf-8 encoded lines with:
+        * word at the start of the line
+        * followed by whitespace
+        * followed by whitespace separated vector components
+
+    Parameters
+    ----------
+    file : str, bytes, int, PathLike
+        Path to a file with embeddings in word2vec binary format.
+
+    Returns
+    -------
+    embeddings : Embeddings
+        Embeddings from the input file. The resulting Embeddings will have a
+        SimpleVocab, NdArray and Norms.
+    """
+    with open(file) as inf:
+        try:
+            first = next(inf)
+        except StopIteration:
+            raise ValueError("Can't read from empty embeddings file.")
+        line = re.split(r'(?a)\s+', first.rstrip())
+        cols = len(line[1:])
+        rows = sum(1 for _ in inf) + 1
+        inf.seek(0)
+        return _load_text(inf, rows, cols)
+
+
+def write_text(file: Union[str, bytes, int, PathLike], embeddings: Embeddings):
+    """
+    Write embeddings in text format.
+
+    Embeddings are un-normalized before serialization, if norms are present, each embedding is
+    scaled by the associated norm.
+
+    The output consists of utf-8 encoded lines with:
+        * word at the start of the line
+        * followed by whitespace
+        * followed by whitespace separated vector components
+
+    Parameters
+    ----------
+    file : str, bytes, int, PathLike
+        Output file
+    embeddings : Embeddings
+        Embeddings to write
+    """
+    _write_text(file, embeddings, False)
+
+
+def write_text_dims(file: Union[str, bytes, int, PathLike],
+                    embeddings: Embeddings):
+    """
+    Write embeddings in text-dims format.
+
+    Embeddings are un-normalized before serialization, if norms are present, each embedding is
+    scaled by the associated norm.
+
+    The output consists of utf-8 encoded lines with:
+        * `rows cols` on the **first** line
+        * word at the start of the line
+        * followed by whitespace
+        * followed by whitespace separated vector components
+
+    Parameters
+    ----------
+    file : str, bytes, int, PathLike
+        Output file
+    embeddings : Embeddings
+        Embeddings to write
+    """
+    _write_text(file, embeddings, True)
+
+
+def _load_text(file: TextIO, rows: int, cols: int) -> Embeddings:
+    words = []
+    matrix = np.zeros((rows, cols), dtype=np.float32)
+    for row, line in zip(matrix, file):
+        parts = re.split(r'(?a)\s+', line.rstrip())
+        words.append(parts[0])
+        row[:] = parts[1:]
```

Comment:
Is htere some explicit conversion here?

Reply:
Conversion seems to happen implicitly and other workarounds seem ugly. `np.fromstring` probably allocates an array for each embedding and going for `list(map(float, parts[1:]))` seems a bit awkward as well.

```Python
>>> import numpy as np
>>> x = np.zeros(5, dtype=np.float32)
>>> x[:] = "1.2"
>>> x
array([1.2, 1.2, 1.2, 1.2, 1.2])
>>> x[:] = "a"
ValueError: could not convert string to float: 'a'
```